test-suite/termstructures.cpp

```diff
@@ -334,6 +335,129 @@ void TermStructureTest::testLinkToNullUnderlying() {
+    rates.push_back(0.117021968693922);
+    rates.push_back(0.109824660896137);
+    rates.push_back(0.109231572878364);
+    rates.push_back(0.119218123236241);
+    rates.push_back(0.128647300167664);
+    rates.push_back(0.0506086995288751);
+
+    boost::shared_ptr<YieldTermStructure> termStructure1 = boost::shared_ptr<YieldTermStructure>(
+        new ForwardCurve(dates, rates, Actual365Fixed(), NullCalendar()));
+
+    // Second curve
+    dates.clear();
+    rates.clear();
+
+    dates.push_back(Date(10, Nov, 2017));
+    dates.push_back(Date(13, Nov, 2017));
+    dates.push_back(Date(11, Dec, 2017));
+    dates.push_back(Date(12, Feb, 2018));
+    dates.push_back(Date(10, May, 2018));
+    dates.push_back(Date(31, Jan, 2022));
+    dates.push_back(Date(7, Dec, 2023));
+    dates.push_back(Date(31, Jan, 2025));
+    dates.push_back(Date(31, Mar, 2028));
+    dates.push_back(Date(7, Dec, 2033));
+    dates.push_back(Date(1, Feb, 2038));
+    dates.push_back(Date(2, Apr, 2046));
+    dates.push_back(Date(2, Jan, 2051));
+    dates.push_back(Date(31, Dec, 2141));
+
+    rates.push_back(0.056656806197189);
+    rates.push_back(0.056656806197189);
+    rates.push_back(0.0419541633454473);
+    rates.push_back(0.0286681050019797);
+    rates.push_back(0.0148840226959593);
+    rates.push_back(0.0246680238374363);
+    rates.push_back(0.0255349067810599);
+    rates.push_back(0.0298907184711927);
+    rates.push_back(0.0263943927922053);
+    rates.push_back(0.0291924526539802);
+    rates.push_back(0.0270049276163556);
+    rates.push_back(0.028775807327614);
+    rates.push_back(0.0293567711641792);
+    rates.push_back(0.010518655099659);
+
+    boost::shared_ptr<YieldTermStructure> termStructure2 = boost::shared_ptr<YieldTermStructure>(
+        new ForwardCurve(dates, rates, Actual365Fixed(), NullCalendar()));
+
+    typedef Real(*binary_f)(Real, Real);
+
+    boost::shared_ptr<YieldTermStructure> compoundCurve = boost::shared_ptr<YieldTermStructure>(
+        new CompositeZeroYieldStructure<binary_f>(Handle<YieldTermStructure>(termStructure1), Handle<YieldTermStructure>(termStructure2), sub));
+
+    // Expected values
+    dates.clear();
+    rates.clear();
+
+    dates.push_back(Date(10, Nov, 2017));
+    dates.push_back(Date(15, Dec, 2017));
+    dates.push_back(Date(15, Jun, 2018));
+    dates.push_back(Date(15, Sep, 2029));
+    dates.push_back(Date(15, Sep, 2038));
+    dates.push_back(Date(15, Mar, 2046));
+    dates.push_back(Date(15, Dec, 2141));
+
+    rates.push_back(0.00892551511527986);
+    rates.push_back(0.0412773974133423);
+    rates.push_back(0.0567251712638837);
+    rates.push_back(0.0878295160422323);
+    rates.push_back(0.0904423159037861);
+    rates.push_back(0.0998714928415959);
+    rates.push_back(0.0400900444382439);
+
+    Real tolerance = 1.0e-10;
+    for (Size i = 0; i < dates.size(); ++i) {
+        Rate actual = compoundCurve->forwardRate(dates[i], dates[i], Actual365Fixed(), Continuous).rate();
```

Comment:
I would check the zero rates, since that's what the curve is composing.  Also, instead of storing the results, this way you can actually check that the resulting zero is the difference of the two original ones.

Reply:
Ah OK, for my use case I do actually want to do operations on the forward rates. I guess I'll need a separate `CompositeForwardTermStructure` ?

Reply:
Yes, but that's more tricky.

In that case, if you only provide `forwardImpl`, the default implementation of `zeroRate(t)` will have to integrate the forwards between 0 and t, which of course is slow and/or inaccurate.  (And unfortunately, the way these classes are implemented, even `forwardRate` goes through zero rates.)

The base class for forward curves also declares `zeroYieldImpl`, so that it can be overridden to return a closed formula for the integral.  But if you pass a generic function, you can't provide such a formula, of course.

So all in all, I think you'll be better off hard-coding the difference function in this case.

Reply:
Ah, I see what you mean. I tried implementing `CompositeForwardTermStructure` and I see the integration. Actually for my case I'm working with continuous rates. So I could use zero rates as well. I was just used to the forward rates, because my input curves are ForwardCurves with BackwardFlat interpolation.